### ISSUE TITLE
Feauture request: Dynamic Lora Weights

### ISSUE THREAD

**Issue by OliviaOliveiira** [comfyanonymous/ComfyUI#2591]
Hey there!
There's an incredibly powerful extension for A1111 called Dynamic Lora Weights which allows one to control the weights of the lora at any given step during the whole generation. For instance - 0.2@0.2,1@0.2 means that lora weight starts at 0.2 till 20% of steps, and then ramps up to 1 from 20% of steps till the end of the generation.
Link to extension - https://github.com/cheald/sd-webui-loractl

**Comment by ltdrdata** [comment]
Take a look at this:
https://github.com/asagi4/comfyui-prompt-control

**Comment by asagi4** [comment]
You can also combine my prompt control utility with any dynamic prompt utility, like the MUWildCard node from my other repo that implements prompt fusion style functions and variables: https://github.com/asagi4/comfyui-utility-nodes
Or you can use the JinjaRender node for even more advanced dynamic prompts.

The syntax isn't quite as nice as sd-webui-loractl, but you can get them to do similar things.

**Comment by brendanhoar** [comment]
IIRC, discussion about this elsewhere indicated that developers were waiting for resolution of https://github.com/comfyanonymous/ComfyUI/pull/2666 , the execution model inversion changes before attempting work like this.

Anyone want to take a look at to see if this is more feasible now?

**Comment by asagi4** [comment]
@tncrdn You don't need JinjaRender for a simple case like that. You can just do `[<lora:A:weight>:<lora:B:weight>:0.25]`. Absolute step counts aren't supported (mostly because it's not available at cond generation time) so you'll have to calculate the fraction, but it'll work. JinjaRender may be useful if you want more complicated schedules since typing those by hand is going to be tedious.

**Comment by tncrdn** [comment]
@asagi4 Thank you very much. And I use PromptToSchedule and ScheduleToModel nodes for that?

**Comment by asagi4** [comment]
@tncrdn yes. PromptToSchedule does the parsing and ScheduleToModel applies a model patch that does the LoRA scheduling.

**Comment by tncrdn** [comment]
@asagi4 Thank you. One last question. Would this be better than using 2 KSampler Advanced nodes and setting start and stop steps?

**Comment by asagi4** [comment]
@tncrdn If you use the PCSplitSampling node to enable split sampling, that's essentially what it will do.

The effects are different though. Doing two ksampler passes isn't quite the same thing as one pass with the same number of steps, at least with some samplers.

**Comment by tncrdn** [comment]
@asagi4 So is this the correct way to use it with SDXL? (I also used a ScheduleToCond) If you can check the attached workflow please?
[workflow.json](https://github.com/user-attachments/files/16911718/workflow.json)

**Comment by asagi4** [comment]
@tncrdn that works, though you don't necessarily need two separate schedules for ScheduleToModel and ScheduleToCond; In fact you'll want to pass the LoRAs into ScheduleToCond too if you want the LoRA to apply to the text encoder, because otherwise they'll only apply to the unet.

**Comment by tncrdn** [comment]
@asagi4 So I wrote the prompt, lora and SDXL(896 1152, 896 1152, 0 0) into one PromptToSchedule node and sent it both to ScheduleToModel and ScheduleToCond. Thank you very much.

**Comment by mcmonkey4eva** [comment]
Native support for dynamic lora weights is being discussed and will likely happen soon(™️ )

**Comment by tncrdn** [comment]
@mcmonkey4eva That's great news. Will it support Flux as well?

**Comment by FilipAndersson245** [comment]
Any news on implementing dynamic lora weights?

**Comment by Veilosity** [comment]
> @tncrdn yes. PromptToSchedule does the parsing and ScheduleToModel applies a model patch that does the LoRA scheduling.

I'm not quite sure about the syntax for changing the Lora strength in Comfyui. Could you please let me know how to have a Lora run at 0 for the first 10 steps, at 1 from 11 to 20, and 0 again from 21 to 30?

**Comment by mcmonkey4eva** [comment]
The branch tracking work towards dynamic lora weights (and related features) is https://github.com/comfyanonymous/ComfyUI/tree/patch_hooks

**Comment by Veilosity** [comment]
I meant the syntax for using the https://github.com/asagi4/comfyui-prompt-control extension.

**Comment by tncrdn** [comment]
> > @tncrdn yes. PromptToSchedule does the parsing and ScheduleToModel applies a model patch that does the LoRA scheduling.
> 
> I'm not quite sure about the syntax for changing the Lora strength in Comfyui. Could you please let me know how to have a Lora run at 0 for the first 10 steps, at 1 from 11 to 20, and 0 again from 21 to 30?

I would try [<lora:somelora:0:1:0>:0.3:0.6:1] but I'm not sure, never tried it before. The right side numbers are fractions for steps 10 of 30, 20 of 30 and 30 of 30. Only fractions work. But like I said I'm not sure if this is the right syntax.

**Comment by WetOnTheWater** [comment]
@mcmonkey4eva is this feature released? Would be great to use

**Comment by mcmonkey4eva** [comment]
yes, announced here https://blog.comfy.org/p/masking-and-scheduling-lora-and-model-weights

**Comment by dylanlacewell** [comment]
@mcmonkey4eva any plans to expose the new lora weights feature natively in Swarm as part of the lora widget that shows up when you click and insert from the catalog?  Maybe a right click to insert a version with extra controls? 

You could argue this is not a "beginner" feature, but many of the concept loras on civit cause unwanted style changes and/or don't have well defined triggers.  Ramping such loras down to a weight of 0 after some fraction of samples helps a lot.  At a minimum, just a hard on/off schedule (no ramp) would probably be useful in the UI.

**Comment by mcmonkey4eva** [comment]
Yes that's planned

**Comment by tncrdn** [comment]
I opened a issue here as well [https://github.com/comfyanonymous/ComfyUI/issues/6628](url) but trying to schedule to 2 loras back to back it's giving me out of memory with 24gb vram. I'm not sure if I'm using it right but here is the workflow.

[workflow7.json](https://github.com/user-attachments/files/18628604/workflow7.json)

![Image](https://github.com/user-attachments/assets/6381418a-9e39-4d2e-9b29-09c016967b79)
